src/names.c

```diff
@@ -0,0 +1,379 @@
+}
+
+static struct sexp_cow as_minimal_names(struct sexp_cow cow_names) {
+  SEXP names = cow_names.obj;
+
+  if (TYPEOF(names) != STRSXP) {
+    Rf_errorcall(R_NilValue, "`names` must be a character vector");
+  }
+
+  R_len_t i = 0;
+  R_len_t n = Rf_length(names);
+  SEXP* ptr = STRING_PTR(names);
+
+  for (; i < n; ++i, ++ptr) {
+    SEXP elt = *ptr;
+    if (elt == NA_STRING) {
+      break;
+    }
+  }
+  if (i == n) {
+    return cow_names;
+  }
+
+  cow_names = r_maybe_copy(cow_names);
+  names = cow_names.obj;
+
+  for (; i < n; ++i, ++ptr) {
+    SEXP elt = *ptr;
+    if (elt == NA_STRING) {
+      SET_STRING_ELT(names, i, vctrs_shared_empty_str_elt);
+    }
+  }
+
+  return cow_names;
+}
+
+SEXP vctrs_as_minimal_names(SEXP names) {
+  struct sexp_cow cow_names = PROTECT_COW(names);
+  cow_names = as_minimal_names(cow_names);
+
+  UNPROTECT(1);
+  return cow_names.obj;
+}
+SEXP vctrs_minimal_names(SEXP x) {
+  SEXP names = PROTECT(vec_bare_names(x));
+
+  if (names == R_NilValue) {
+    UNPROTECT(1);
+    return Rf_allocVector(STRSXP, vec_size(x));
+  }
+
+  struct sexp_cow cow_names = PROTECT_COW(names);
+  cow_names = as_minimal_names(cow_names);
+
+  UNPROTECT(2);
+  return cow_names.obj;
+}
+
+
+void stop_large_name();
+bool is_dotdotint(const char* name);
+ptrdiff_t suffix_pos(const char* name);
+
+static struct sexp_cow as_unique_names(struct sexp_cow cow_names) {
+  SEXP names = cow_names.obj;
+
+  if (TYPEOF(names) != STRSXP) {
+    Rf_errorcall(R_NilValue, "`names` must be a character vector");
+  }
+
+  dictionary d;
+  dict_init(&d, names);
+  SEXP dups = PROTECT(Rf_allocVector(INTSXP, d.size));
+  int* dups_ptr = INTEGER(dups);
+
+  R_len_t i = 0;
+  R_len_t n = Rf_length(names);
+  SEXP* ptr = STRING_PTR(names);
+
+  // First quick pass to detect if any repairs are needed. See second
+  // part of the loop for the meaning of each branch.
+  for (; i < n; ++i, ++ptr) {
+    SEXP elt = *ptr;
+
+    if (elt == NA_STRING || elt == strings_dots || elt == strings_empty || is_dotdotint(CHAR(elt))) {
+      break;
+    }
+    if (suffix_pos(CHAR(elt)) >= 0) {
+      break;
+    }
+
+    int32_t k = dict_find(&d, names, i);
+
+    if (d.key[k] == DICT_EMPTY) {
+      dict_put(&d, k, i);
+      dups_ptr[k] = 1;
+    } else {
+      break;
+    }
+  }
+
+  // Return early when no repairs are needed
+  if (i == n) {
+    UNPROTECT(1);
+    return cow_names;
+  }
+
+
+  cow_names = r_maybe_copy(cow_names);
+  names = cow_names.obj;
+  ptr = STRING_PTR(names) + i;
+
+  for (; i < n; ++i, ++ptr) {
+    SEXP elt = *ptr;
+
+    // Set `NA` and dots values to "" so they get replaced by `...n` later on
+    if (elt == NA_STRING || elt == strings_dots || is_dotdotint(CHAR(elt))) {
+      elt = strings_empty;
+      SET_STRING_ELT(names, i, elt);
+      SET_STRING_ELT(d.x, i, elt);
+    }
+
+    // Strip `...n` suffixes
+    const char* nm = CHAR(elt);
+    int pos = suffix_pos(nm);
+    if (pos >= 0) {
+      R_CheckStack2(pos + 1);
```

Comment:
What does this do?

Reply:
This checks there is sufficient space to allocate on the stack.

Reply:
Oh `buf[pos + 1]` is doing the allocation. Isn't that a C99 idiom? I thought we couldn't use that in R code.